clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -643,7 +639,9 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                     error);
                 if (transactionManager == null) {
                     reenqueueBatch(batch, now);
-                } else if (transactionManager.hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch())) {
+                } else if (transactionManager.hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch()) ||
```

Comment:
There's one case where we pass in the batch's epoch + 1, and I couldn't think of a clean way to improve it. Let me know if you feel strongly and I'll take another look

Reply:
Just &&'ing with the RetriableException check doesn't quite work because we want to retry OutOfOrderSequence errors in some cases, but I tried to move most of the logic inside the TransactionManager. It allows us to remove the ID and epoch check in Sender.